ideas/000-discoverable-trusted-server-nodes/README.md

```diff
@@ -0,0 +1,73 @@
+---
+id: 000-discoverable-trusted-server-nodes
+title: Discoverable And Trusted Server Nodes
+status: Draft
+created: 2018-06-07
+category: core
+lead-contributor: adambabik
+contributors:
+exit-criteria: yes
+success-metrics: yes
+clear-roles: no
+future-iterations: yes
+roles-needed:
+    - Smart Contract Developer
+    - Go Developer
+    - Clojure Developer
+    - UI/UX Designer
+okrs:
+    - "[PO]: Objective: Messaging is reliable"
+    - "[P2]: KR: Cluster can handle 500 concurrent users"
+---
+
+## Summary
+The idea here is to provide a way to make Mail Server nodes discoverable and trusted. This should work seamlessly for the end user.
+
+## Product Overview
+Currently, Mail Servers (and other planned server nodes like Push Notification servers) need to be hardcoded in the app code. It's possible to provide your own Mail Server in the app settings but that's not a scalable solution and it does not solve all issues.
+
+The outcome of this swarm are fully decentralized and discoverable Mail Servers. The Status app will have a mean to verify that a discovered Mail Server is trusted.
+
+In the next iterations, we may work on a way to introduce external providers to submit their own Mail Servers into the cluster and allow end users to accept these resources as trusted.
+
+## Product Description
+The work in this swarm can be divided into three areas:
+- (Infra) making Mail servers discoverable,
+- (Smart Contract) writing a smart contract to keep track of trusted Mail Servers,
+- (Client) mechanism of verifying whether a discovered Mail Server is trusted.
+
+As mentioned in the Product Overview, the smart contract should allow registering custom providers. The client will have an option to decide which providers it trust. Status will be a default provider and by default the app will trust it.
+
+Due to the fact that all Mail Servers are independent and keep the same history of messages, we will be able to remove all hardcoded Mail Server addresses from the client and rely only on discovered Mail Servers. This will allow Status to dynamically scale number of their the Mail Servers depending on the usage.
+
+**Note** Important thing is that this solution should be flexible for other server nodes. For example, consider Push Notification servers that should be discovered and trusted in the same way.
```

Comment:
last time i checked there was no such requirement for push notifications, as they don't need direct p2p connection. but in general i don't argue that this should be "flexible"

Reply:
@adriacidre, do you know more about it? Is it a valid case or Push Notification servers will be advertised differently (Whisper?)?

Reply:
Ok, thanks for the explanation. I will remove Push Notification servers from the examples.

Reply:
To me push notifications server and mail server should be the same, encapsulated in a 'masternode' service, possibly with different capabilities (PN/mailserver).

Reply:
I am not familiar with the masternode concept and PN serves but the point here is that this solution is going to be flexible and verification will be put on the client's shoulders. As long as a particular group of servers is discoverable and the client can choose arbitrary one, it will work.

Reply:
We started to work on push notification servers discovery process fully based on whisper, however, I think we should be able to adapt it and re-use most of the parts of this approach.

cc: @PombeirP

Reply:
i noted somewhere that the benefit of doing so will be that initial topic will be unknown, and thus nobody will be able to spam it.

but on the other hand if there is no need for direct p2p connection i don't see any good argument to enforce it, besides convenience

Reply:
I definitely agree that we don't want or need two different ways of discovering nodes, so I'm happy to reuse the approach that we agree on here for the future PN server work.